### PR TITLE
[SPARK-44169][SQL] Assign names to the error class _LEGACY_ERROR_TEMP_[2300-2304]

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The pr aims to assign names to the error class _LEGACY_ERROR_TEMP_[2300-2304].

### Why are the changes needed?
Improve the error framework.

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
Exists test cases updated and added new test cases.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @MaxGekk

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -1209,6 +1226,23 @@
...
+    ],
+    "subClass" : {
+      "ARGUMENTS_NUMBER_MISMATCH" : {
+        "message" : [
+          "The number of lambda function arguments '<namesSize>' does not match the number of arguments expected by the higher order function '<argInfoSize>'."
+        ]
+      },
+      "ARGUMENT_NAMES_EXISTS_SAME_SEMANTIC" : {
+        "message" : [
+          "Lambda function arguments should not have names that are semantically the same."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1241]: Could you make the error message clear and more precise. From the implementation:
```
    if (!conf.caseSensitiveAnalysis) {
      // scalastyle:off caselocale
      s: String => s.toLowerCase
      // scalastyle:on caselocale
    } else {
      s: String => s
    }
``` 

Like:
```
A lambda function has duplicate arguments: <dupArgs>. Consider to rename the arguments or to set the SQL config <caseSensitiveConfig> to \"false\".
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 2]
```diff
@@ -5189,6 +5189,18 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
         stop = 22)
     )
 
+    checkError(
+      exception = intercept[AnalysisException] {
+        df.selectExpr("zip_with(a1, a2, (x, x) -> x)")
+      },
+      errorClass = "INVALID_LAMBDA_FUNCTION_CALL.ARGUMENT_NAMES_EXISTS_SAME_SEMANTIC",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:5196]: The error class name confuses slightly. Let's improve it.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala [hunk 3]
```diff
@@ -5179,7 +5179,7 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
       exception = intercept[AnalysisException] {
         df.selectExpr("zip_with(a1, a2, x -> x)")
       },
-      errorClass = "_LEGACY_ERROR_TEMP_2300",
+      errorClass = "INVALID_LAMBDA_FUNCTION_CALL.ARGUMENTS_NUMBER_MISMATCH",
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala:5182]: Let's make it consistent to already existing error classes:
- NUM_COLUMNS_MISMATCH
- WRONG_NUM_ARGS

and make it shorter slightly: `NUM_ARGS_MISMATCH`

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -1654,6 +1688,11 @@
     ],
     "sqlState" : "42803"
   },
+  "MULTI_SOURCES_UNSUPPORTED_FOR_EXPRESSION" : {
+    "message" : [
+      "'<name>' does not support more than one sources."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1693]:
```suggestion
      "The expression <expr> does not support more than one source."
```

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -1209,6 +1226,23 @@
...
     "sqlState" : "22032"
   },
+  "INVALID_LAMBDA_FUNCTION_CALL" : {
+    "message" : [
+      "Invalid lambda function call."
+    ],
+    "subClass" : {
+      "ARGUMENTS_NUMBER_MISMATCH" : {
+        "message" : [
+          "The number of lambda function arguments '<namesSize>' does not match the number of arguments expected by the higher order function '<argInfoSize>'."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1236]:
```suggestion
          "A higher order function expects <expectedNumArgs> arguments, but got <actualNumArgs>."
```

File: core/src/main/resources/error/error-classes.json [hunk 6]
```diff
@@ -1197,6 +1192,28 @@
...
     "sqlState" : "22003"
   },
+  "INVALID_INLINE_TABLE" : {
+    "message" : [
+      "Invalid inline table."
+    ],
+    "subClass" : {
+      "CANNOT_EVALUATE_EXPRESSION_IN_INLINE_TABLE" : {
+        "message" : [
+          "cannot evaluate expression <sqlExpr> in inline table definition."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1202]:
```suggestion
          "Cannot evaluate the expression <expr> in inline table definition."
```

File: core/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -1197,6 +1192,28 @@
...
+        ]
+      },
+      "FAILED_SQL_EXPRESSION_EVALUATION" : {
+        "message" : [
+          "Failed to evaluate the SQL expression: <sqlExpr>. Please check your syntax and ensure all required tables and columns are available."
+        ]
+      },
+      "INCOMPATIBLE_TYPES_IN_INLINE_TABLE" : {
+        "message" : [
+          "incompatible types found in column <name> for inline table."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1212]:
```suggestion
          "Incompatible types found in column <name> for inline table."
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (beliefer) [reply core/src/main/resources/error/error-classes.json:1241]: I don't get your point yet.
I guess you mean is `to set the SQL config <caseSensitiveConfig> to \"true\"`.

**Comment by beliefer** [comment]
The failure GA is unrelated, let's re-trigger it.

File: core/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -1209,6 +1226,23 @@
...
     "sqlState" : "22032"
   },
+  "INVALID_LAMBDA_FUNCTION_CALL" : {
+    "message" : [
+      "Invalid lambda function call."
+    ],
+    "subClass" : {
+      "ARGUMENT_NAMES_EXISTS_SAME_SEMANTIC" : {
+        "message" : [
+          "Lambda function arguments should not have names that are semantically the same."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1236]: From my point of view, the message is overcomplicated. What does `semantically` mean?
Let's make it more precise:
```suggestion
          "The lambda function has duplicate arguments <args>. Please, consider to rename the argument names or set <caseSensitiveConfig> to \"true\"."
```

File: core/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -1197,6 +1192,28 @@
...
+    ],
+    "subClass" : {
+      "CANNOT_EVALUATE_EXPRESSION_IN_INLINE_TABLE" : {
+        "message" : [
+          "Cannot evaluate the expression <expr> in inline table definition."
+        ]
+      },
+      "FAILED_SQL_EXPRESSION_EVALUATION" : {
+        "message" : [
+          "Failed to evaluate the SQL expression: <sqlExpr>. Please check your syntax and ensure all required tables and columns are available."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1207]:
```suggestion
          "Failed to evaluate the SQL expression <sqlExpr>. Please check your syntax, and ensure all required tables and columns are available."
```

File: core/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -1197,6 +1192,28 @@
...
+        ]
+      },
+      "INCOMPATIBLE_TYPES_IN_INLINE_TABLE" : {
+        "message" : [
+          "Incompatible types found in column <name> for inline table."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1212]:
```suggestion
          "Found incompatible types in the column <colName> for inline table."
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala [hunk 11]
```diff
@@ -96,7 +96,7 @@ object ResolveInlineTables extends Rule[LogicalPlan] with CastSupport with Alias
       val inputTypes = column.map(_.dataType)
       val tpe = TypeCoercion.findWiderTypeWithoutStringPromotion(inputTypes).getOrElse {
         table.failAnalysis(
-          errorClass = "_LEGACY_ERROR_TEMP_2303",
+          errorClass = "INVALID_INLINE_TABLE.INCOMPATIBLE_TYPES_IN_INLINE_TABLE",
           messageParameters = Map("name" -> name))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveInlineTables.scala:100]: Is `name` an identifier? If so, wrap it by `toSQLId()`.

File: core/src/main/resources/error/error-classes.json [hunk 12]
```diff
@@ -1209,6 +1226,23 @@
     ],
     "sqlState" : "22032"
   },
+  "INVALID_LAMBDA_FUNCTION_CALL" : {
+    "message" : [
+      "Invalid lambda function call."
+    ],
+    "subClass" : {
+      "ARGUMENT_NAMES_EXISTS_SAME_SEMANTIC" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1234]:
```suggestion
      "DUPLICATE_ARG_NAMES" : {
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala [hunk 13]
```diff
@@ -72,16 +72,18 @@ object ResolveLambdaVariables extends Rule[LogicalPlan] {
...
       }
 
       if (names.map(a => canonicalizer(a.name)).distinct.size < names.size) {
         e.failAnalysis(
-          errorClass = "_LEGACY_ERROR_TEMP_2301",
-          messageParameters = Map.empty)
+          errorClass = "INVALID_LAMBDA_FUNCTION_CALL.DUPLICATE_ARG_NAMES",
+          messageParameters = Map(
+            "args" -> names.map(a => canonicalizer(a.name)).mkString(", "),
+            "caseSensitiveConfig" -> "spark.sql.caseSensitive"))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:86]: Could you wrap the SQL config by `toSQLConf` like:
```suggestion
            "caseSensitiveConfig" -> toSQLConf(SQLConf.CASE_SENSITIVE.key)))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala [hunk 14]
```diff
@@ -72,16 +72,18 @@ object ResolveLambdaVariables extends Rule[LogicalPlan] {
...
+            "actualNumArgs" -> argInfo.size.toString))
       }
 
       if (names.map(a => canonicalizer(a.name)).distinct.size < names.size) {
         e.failAnalysis(
-          errorClass = "_LEGACY_ERROR_TEMP_2301",
-          messageParameters = Map.empty)
+          errorClass = "INVALID_LAMBDA_FUNCTION_CALL.DUPLICATE_ARG_NAMES",
+          messageParameters = Map(
+            "args" -> names.map(a => canonicalizer(a.name)).mkString(", "),
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:85]: Is `name` an id? If so, wrap it by `toSQLId`

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala (hunk 13 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/higherOrderFunctions.scala:86]: Got it.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by beliefer** [comment]
@MaxGekk GA passed.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @beliefer.

**Comment by beliefer** [comment]
@MaxGekk Thank you for the hard work.

**Pull request closed without merging**